### ISSUE TITLE
Rule proposal: no-wildcard

### ISSUE THREAD

**Issue by silvenon** [benmosher/eslint-plugin-import#1011]
I would like to propose a `no-wildcard` rule, which would prohibit the following:

```js
import * as foo from './foo';
export * as foo from './foo';
```

I'm aware that the latter is an experimental export extension, but implementing it probably has no cost other than adding another selector.

One of the reasons for this rule to exist is because wildcard imports ruin tree-shaking in module bundlers, often without the developer's knowledge.

The rule could also have an option `ignore`/`exclude`/`excludedModules`/etc. which would be a list of modules where this rule would allow wildcard imports. For example, Flow docs [recommend wildcard import for React](https://flow.org/en/docs/react/components/) in order to import React's utility types, so in that case it would make sense to define the rule as something like:

```yml
import/no-wildcard:
  - error
  - ignore:
    - react
```

This rule was initially proposed in eslint/eslint#4865, but didn't receive much interest from the core team (should we ping them and wait longer?). Maybe if `eslint-plugin-import` popularizes this good practice, the `eslint` team would be more keen to implement it in the core.

**Comment by ljharb** [comment]
I think that there’s a case for having the rule here, but having it in core is ideal; and in the meantime, using no-restricted-syntax can suffice.

**Comment by TrevorBurnham** [comment]
@silvenon Doesn't the existing [`no-namespace`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-namespace.md) rule do what you're requesting here? (Aside from the `export *` part, I mean.)

**Comment by ljharb** [comment]
Also, using the new form doesn’t impact treeshaking - it’s the import * as that does that.

**Comment by silvenon** [comment]
That's true! I think we can close this then.
